modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/start_stop_client/node/ActionNode.java

```diff
@@ -0,0 +1,152 @@
+     * */
+    private ConcurrentLinkedDeque<Report> queue = new ConcurrentLinkedDeque<Report>();
+
+    /* **/
+    Reporter reporter;
+
+    /* **/
+    Metadata testInfo = new Metadata();
+
+    /** Start test time. */
+    protected Date startTime;
+
+    /** End test time. */
+    protected Date endTime;
+
+    /** Thread count. */
+    private int threads;
+
+    /** The delay between iterations of the action. */
+    protected long pacing;
+
+    /** Current action name. */
+    protected String actionName;
+
+    /** Thread executer. */
+    protected ExecutorService executor;
+
+    /** Threads. */
+    private ArrayList<OperationThread> operationThreads = new ArrayList();
+
+    /**  */
+    private Logger log = LogManager.getLogger(ActionNode.class.getName());
+
+    /** {@inheritDoc} */
+    @Override protected void run(JsonNode jsonNode) throws Exception {
+        init(jsonNode);
+
+        markInitialized();
+
+        for (int i = 0; i < threads; i++) {
+            operationThreads.add(new OperationThread(this, log, pacing));
+        }
+        for (OperationThread thread : operationThreads) {
+            executor.execute(thread);
+        }
+
+        while (!terminated()) {
+            TimeUnit.MILLISECONDS.sleep(100);
+        }
+
+        executor.shutdownNow();
+        if (!executor.isShutdown()) {
+            TimeUnit.MILLISECONDS.sleep(100);
+        }
+
+        endTime = new Date();
+        markFinished();
+    }
+
+    /** */
+    protected void init(JsonNode jsonNode) throws IOException {
+        threads = Optional.ofNullable(jsonNode.get("threads")).map(JsonNode::asInt).orElse(1);
+        pacing = Optional.ofNullable(jsonNode.get("pacing")).map(JsonNode::asLong).orElse(0l);
+        actionName = Optional.ofNullable(jsonNode.get("action")).map(JsonNode::asText).orElse("default-action-name");
+        String reportFolder = jsonNode.get("report_folder").asText();
+
+        startTime = new Date();
+
+        log.info(
+                "test props:" +
+                " action=" + actionName +
+                " pacing=" + pacing +
+                " threads=" + threads
+        );
+
+        testInfo.setActionName(actionName);
+        testInfo.setThreads(threads);
+        testInfo.setStartTime(System.currentTimeMillis());
+
+        File folder = new File(reportFolder);
+        if (!folder.exists()) {
+            folder.mkdir();
+        }
+
+        String reportPath = reportFolder + "/report-" + System.currentTimeMillis() + ".csv";
+        reporter = new Reporter(reportPath, queue, testInfo);
+
+        executor = Executors.newFixedThreadPool(threads + 3);
+        executor.execute(reporter);
+        scriptInit(jsonNode);
+    }
+
+    /** Init method. */
+    protected abstract void scriptInit(JsonNode jsonNode);
+
+    /** {@inheritDoc} */
+    @Override public void publishInternalReport(Report report) {
+        if (!executor.isShutdown()) {
+            executor.execute(new Runnable() {
```

Comment:
Why you create new instance of Runnable in queue(execute in executor) to push something to queue?
Looks super weird.